File: web-ide/frontend/src/types/file.ts
import { Cell } from './cell';

export interface FileLeaf {
  id: number;
  name: string;
  type: 'file';
  project_id: number;
  folder_id: number;
  content: Content;
}

export interface FolderNode {
  folderId: number;
  folderName: string;
  parentId: number | null;
  projectId: number;
  files?: FileLeaf[];
  children?: FolderNode[];
}

export interface Content {
  cells: Cell[];
}

// export interface File {
//   id: number;
//   name: string;
//   content: string[];
// }

// export interface Folder {
//   folderId: number;
//   folderName: string;
//   files: File[];
//   folders: Folder[];
// }

File: web-ide/backend/src/main/java/io/goorm/backend/entity/File.java
package io.goorm.backend.entity;

import com.fasterxml.jackson.annotation.JsonBackReference;
import com.fasterxml.jackson.annotation.JsonIgnore;
import jakarta.persistence.*;
import lombok.*;
import org.hibernate.annotations.CreationTimestamp;
import org.hibernate.annotations.UpdateTimestamp;

import java.time.LocalDateTime;

@Entity
@Table(name = "files")
@Getter
@NoArgsConstructor(access = AccessLevel.PROTECTED)
public class File {

    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @Setter
    @Column(nullable = false, length = 255)
    private String name;

    @Column(columnDefinition = "TEXT")
    private String content;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "folder_id", nullable = false)
    @JsonBackReference
    private Folder folder;

    @ManyToOne(fetch =  FetchType.LAZY)
    @JoinColumn(name = "project_id", nullable = false)
    @JsonIgnore
    private Project project;

    @CreationTimestamp
    @Column(name = "created_at", updatable = false)
    private LocalDateTime createdAt;

    @UpdateTimestamp
    @Column(name = "updated_at")
    private LocalDateTime updatedAt;

    @Builder
    public File(String name, String content, Folder folder, Project project) {
        this.name = name;
        this.content = content;
        this.folder = folder;
        this.project = project;
    }

}